File: backend/tests/chain/transactions.test.ts
/**
 * Tests for chain/transactions.ts — listTransactions pulls logs from
 * the last N blocks and enriches them with the underlying transaction
 * envelope (gas price, input, nonce, etc.) and a TxType classification.
 */
import { describe, it, expect, vi } from 'vitest';
import type { Log, TransactionResponse } from 'ethers';
import { listTransactions } from '../../src/chain/transactions.js';
import { TOPIC_V2_SWAP, TOPIC_V3_SWAP } from '../../src/chain/classify.js';

function makeProvider(opts: {
  blockNumber: number;
  logs: Log[];
  txByHash: Map<string, TransactionResponse | null>;
  blockTimestamps?: Map<number, number>;
}) {
  const blockTimestamps = opts.blockTimestamps ?? new Map<number, number>([[100, 1700000000]]);
  return {
    getBlockNumber: vi.fn(async () => opts.blockNumber),
    getLogs: vi.fn(async (filter: { fromBlock: number; toBlock: number; topics?: unknown[] }) => {
      // Naive: return opts.logs if filter range covers the log block.
      return opts.logs.filter(
        (l) => l.blockNumber >= filter.fromBlock && l.blockNumber <= filter.toBlock,
      );
    }),
    getTransaction: vi.fn(async (hash: string) => {
      return opts.txByHash.get(hash) ?? null;
    }),
    getBlock: vi.fn(async (tag: number) => ({
      number: tag,
      hash: '0x' + '00'.repeat(32),
      parentHash: '0x' + '00'.repeat(32),
      timestamp: blockTimestamps.get(tag) ?? 1700000000,
      gasLimit: 0n,
      gasUsed: 0n,
      miner: '0x' + '00'.repeat(20),
      transactions: [],
    })),
  };
}

const AAVE_POOL = '0x87870Bca3F3fD6335C3F4ce8392D69350B4fA4E2';
const V2_POOL = '0xB4e16d0168e52d35CaCD2c6185b44281Ec28C9Dc';
const V3_POOL = '0x88e6A0c2dDD26FEEb64F039a2c41296FcB3f5640';
const TX_LIQ = '0x' + 'a1'.repeat(32);
const TX_V2 = '0x' + 'a2'.repeat(32);
const TX_V3 = '0x' + 'a3'.repeat(32);
const TOPIC_AAVE_LIQ = '0x1bfeb1aef4b3e33145cffddd62ce41edf6038b7b0372e76737c498bf28073c1a';

function makeLog(overrides: Partial<Log> = {}): Log {
  return {
    address: V3_POOL,
    blockHash: '0x' + 'bb'.repeat(32),
    blockNumber: 100,
    data: '0x',
    topics: [TOPIC_V3_SWAP],
    transactionHash: TX_V3,
    transactionIndex: 0,
    index: 0,
    removed: false,
    ...overrides,
  } as Log;
}

function makeTx(hash: string, from: string, to: string): TransactionResponse {
  return {
    hash,
    blockHash: '0x' + '00'.repeat(32),
    blockNumber: 100,
    from,
    to,
    value: 0n,
    gasLimit: 200000n,
    gasPrice: 50_000_000_000n,
    nonce: 7,
    input: '0xabcd',
    type: 2,
  } as unknown as TransactionResponse;
}

describe('listTransactions', () => {
  it('returns empty array when no logs match', async () => {
    const provider = makeProvider({ blockNumber: 100, logs: [], txByHash: new Map() });
    const txs = await listTransactions(provider as never, { blocks: 5, limit: 100 });
    expect(txs).toEqual([]);
  });

  it('deduplicates logs by transaction hash (same tx, two logs → 1 tx)', async () => {
    const tx = makeTx(TX_V3, '0x' + '11'.repeat(20), V3_POOL);
    const logs: Log[] = [
      makeLog({ topics: [TOPIC_V3_SWAP], transactionHash: TX_V3, index: 0, blockNumber: 100 }),
      makeLog({ topics: [TOPIC_V3_SWAP], transactionHash: TX_V3, index: 1, blockNumber: 100 }),
    ];
    const provider = makeProvider({
      blockNumber: 100,
      logs,
      txByHash: new Map([[TX_V3, tx]]),
    });
    const txs = await listTransactions(provider as never, { blocks: 5, limit: 100 });
    expect(txs).toHaveLength(1);
    expect(txs[0].hash).toBe(TX_V3);
    expect(provider.getTransaction).toHaveBeenCalledTimes(1);
  });

  it('returns mixed V2 swap, V3 swap, and Aave liquidation in correct shape', async () => {
    const liqTx = makeTx(TX_LIQ, '0x' + '88'.repeat(20), AAVE_POOL);
    const v2Tx = makeTx(TX_V2, '0x' + '77'.repeat(20), V2_POOL);
    const v3Tx = makeTx(TX_V3, '0x' + '66'.repeat(20), V3_POOL);
    const logs: Log[] = [
      makeLog({ address: AAVE_POOL, topics: [TOPIC_AAVE_LIQ], transactionHash: TX_LIQ, blockNumber: 100 }),
      makeLog({ address: V2_POOL, topics: [TOPIC_V2_SWAP], transactionHash: TX_V2, blockNumber: 100 }),
      makeLog({ address: V3_POOL, topics: [TOPIC_V3_SWAP], transactionHash: TX_V3, blockNumber: 100 }),
    ];
    const provider = makeProvider({
      blockNumber: 100,
      logs,
      txByHash: new Map([
        [TX_LIQ, liqTx],
        [TX_V2, v2Tx],
        [TX_V3, v3Tx],
      ]),
    });
    const txs = await listTransactions(provider as never, { blocks: 10, limit: 100 });
    expect(txs).toHaveLength(3);
    // Shape check
    for (const t of txs) {
      expect(typeof t.hash).toBe('string');
      expect(typeof t.from).toBe('string');
      expect(typeof t.gasPrice).toBe('bigint');
      expect(typeof t.gasLimit).toBe('bigint');
      expect(typeof t.nonce).toBe('number');
      expect(typeof t.timestamp).toBe('number');
      expect(typeof t.type).toBe('string');
    }
    // Classification
    const types = new Set(txs.map((t) => t.type));
    expect(types.has('liquidation')).toBe(true);
    expect(types.has('normal')).toBe(true);
  });

  it('respects the `limit` option and caps results', async () => {
    const hashes = Array.from({ length: 5 }, (_, i) => '0x' + (i + 1).toString(16).padStart(64, '0'));
    const logs: Log[] = hashes.map((h, i) =>
      makeLog({ topics: [TOPIC_V3_SWAP], transactionHash: h, index: i }),
    );
    const txByHash = new Map<string, TransactionResponse | null>();
    for (const h of hashes) {
      txByHash.set(h, makeTx(h, '0x' + '11'.repeat(20), V3_POOL));
    }
    const provider = makeProvider({ blockNumber: 100, logs, txByHash });
    const txs = await listTransactions(provider as never, { blocks: 5, limit: 3 });
    expect(txs).toHaveLength(3);
  });

  it('skips transactions that the provider cannot fetch (null tx response)', async () => {
    const logs: Log[] = [
      makeLog({ topics: [TOPIC_V3_SWAP], transactionHash: TX_V3 }),
    ];
    const provider = makeProvider({
      blockNumber: 100,
      logs,
      txByHash: new Map([[TX_V3, null]]), // simulate pruned / unknown tx
    });
    const txs = await listTransactions(provider as never, { blocks: 5, limit: 100 });
    expect(txs).toEqual([]);
  });

  it('queries a single block when blocks=1', async () => {
    const provider = makeProvider({ blockNumber: 50, logs: [], txByHash: new Map() });
    await listTransactions(provider as never, { blocks: 1, limit: 10 });
    expect(provider.getLogs).toHaveBeenCalledTimes(1);
    const arg = (provider.getLogs.mock.calls[0]![0] as { fromBlock: number; toBlock: number });
    expect(arg.toBlock - arg.fromBlock + 1).toBe(1);
  });

  it('returns Transaction objects that include timestamp and blockNumber', async () => {
    const tx = makeTx(TX_V3, '0x' + '11'.repeat(20), V3_POOL);
    const logs: Log[] = [makeLog({ topics: [TOPIC_V3_SWAP], transactionHash: TX_V3 })];
    const provider = makeProvider({
      blockNumber: 100,
      logs,
      txByHash: new Map([[TX_V3, tx]]),
    });
    const [t] = await listTransactions(provider as never, { blocks: 5, limit: 10 });
    expect(t.blockNumber).toBe(100);
    expect(t.timestamp).toBeGreaterThan(0);
  });
});

File: backend/tests/chain/transactions_test.py
"""Unit tests for `dtm_backend.chain.transactions`."""
from __future__ import annotations

import pytest

from dtm_backend.chain.classify import LIQUIDATION_TOPIC
from dtm_backend.chain.transactions import (
    TransactionFetcher,
)
from dtm_backend.chain.types import MevType


def _stub_call(function_name, *args):
    async def _inner():
        # getLogs returns 1 V2 swap, 1 V3 swap, 1 liquidation.
        if function_name == "getLogs":
            return [
                {
                    "address": "0x" + "1" * 40,
                    "blockHash": "0x" + "a" * 64,
                    "blockNumber": 0x10,
                    "data": "0x",
                    "logIndex": "0x0",
                    "topics": ["0x" + "b" * 64],
                    "transactionHash": "0x" + "c" * 64,
                    "transactionIndex": "0x0",
                    "removed": False,
                },
                {
                    "address": "0x" + "2" * 40,
                    "blockHash": "0x" + "d" * 64,
                    "blockNumber": 0x11,
                    "data": "0x",
                    "logIndex": "0x1",
                    "topics": ["0x" + "e" * 64],
                    "transactionHash": "0x" + "f" * 64,
                    "transactionIndex": "0x0",
                    "removed": False,
                },
                {
                    "address": "0x" + "3" * 40,
                    "blockHash": "0x" + "1" * 64,
                    "blockNumber": 0x12,
                    "data": "0x",
                    "logIndex": "0x2",
                    "topics": [LIQUIDATION_TOPIC],
                    "transactionHash": "0x" + "2" * 64,
                    "transactionIndex": "0x0",
                    "removed": False,
                },
            ]
        if function_name == "getTransaction":
            tx_hash = args[0] if args else "0x" + "9" * 64
            return {
                "hash": tx_hash,
                "from": "0x" + "8" * 40,
                "to": "0x" + "1" * 40,
                "value": 0,
                "gasPrice": 10**9,
                "gas": 200_000,
                "input": "0x",
                "nonce": 1,
                "blockNumber": 0x10,
            }
        if function_name == "getBlock":
            return {"timestamp": 1_700_000_000}
        raise AssertionError(f"unexpected call {function_name}")
    return _inner()


@pytest.mark.asyncio
async def test_transaction_fetcher_returns_canonical_three_txs() -> None:
    """`list_transactions` returns ≥1 transaction with the right shape."""
    f = TransactionFetcher(_call=_stub_call)
    txs = await f.list_recent()
    assert len(txs) >= 1
    for t in txs:
        # Wire-format invariants.
        assert isinstance(t.hash, str)
        assert isinstance(t.sender, str)
        assert isinstance(t.timestamp, int)
        assert isinstance(t.gas_limit, int)
        assert isinstance(t.gas_price, int)
        assert t.mev_type in MevType


@pytest.mark.asyncio
async def test_transaction_fetcher_classifies_liquidation() -> None:
    """A log with the LiquidationCall topic produces a LIQUIDATION tx."""
    f = TransactionFetcher(_call=_stub_call)
    txs = await f.list_recent()
    liquidations = [t for t in txs if t.mev_type is MevType.LIQUIDATION]
    assert len(liquidations) >= 1


def test_transaction_fetcher_handles_empty_logs() -> None:
    """`list_recent` returns [] when `getLogs` returns no entries."""

    def _empty_call(fn, *args):
        async def _inner():
            if fn == "getLogs":
                return []
            raise AssertionError(f"unexpected call {fn}")
        return _inner()

    f = TransactionFetcher(_call=_empty_call)
    import asyncio
    assert asyncio.run(f.list_recent()) == []


def test_transaction_fetcher_handles_rpc_error() -> None:
    """`list_recent` propagates network errors."""

    def _err_call(fn, *args):
        async def _inner():
            raise ConnectionError("rpc down")
        return _inner()

    f = TransactionFetcher(_call=_err_call)
    import asyncio
    with pytest.raises(ConnectionError):
        asyncio.run(f.list_recent())
